Reorganize inflate window layout

nmoinvaz opened the pull request:
This commit merges in the Intel inflate changes for crc32 pclmulqdq:
https://github.com/jtkukunas/zlib/commit/62b7616b0bad4ff088f1a73d0fa11159b6925fbb

Here is Jim Kukunas original notes from the commit:

> This commit significantly improves inflate performance by reorganizing the window buffer into a contiguous window and pending output buffer. The goal of this layout is to reduce branching, improve cache locality, and enable for the use of crc folding with gzip input.

> The window buffer is allocated as a multiple of the user-selected window size. In this commit, a factor of 2 is utilized.

> The layout of the window buffer is divided into two sections. The first section, window offset [0, wsize), is reserved for history that has already been output. The second section, window offset [wsize, 2 * wsize), is reserved for buffering pending output that hasn't been flushed to the user's output buffer yet.

> The history section grows downwards, towards the window offset of 0. The pending output section grows upwards, towards the end of the buffer. As a result, all of the possible distance/length data that may need to be copied is contiguous. This removes the need to stitch together output from 2 separate buffers.

> In the case of gzip input, crc folding is used to copy the pending output to the user's buffers.

The downsides:

* inflate uses ~~64kb of memory~~ twice as much window memory. If window is 32k, it uses 64k.
* ~~inflateBack no longer uses inflate_fast. With inflateBack you specify the window buffer and size which is different than what this commit does in that it uses a single memory allocation for window buffer and pending output buffer similar to deflate. We could possibly use a separate inflate_fast function or with separate parameters for inflateBack.~~
* For inflateBack we had to create a separate `zng_inflate_fast_back` function to preserve speed improvements when all we are supplied with is a buffer allocation only intended for use as the window.

Todo:

* ~~Make all CI work~~
* ~~Newer benchmarks~~
* ~~See if we can overcome inflateBack issue.~~

Old Benchmarks:

### ZLIB-NG
```
 Tool: minigzip-ng-develop.exe Levels: 1-9
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     44.809%      1.015/1.022/1.030/0.005        0.660/0.670/0.674/0.003      104,067,822
 2     35.391%      1.678/1.700/1.710/0.009        0.673/0.686/0.692/0.005       82,195,416
 3     33.698%      2.320/2.338/2.348/0.009        0.655/0.661/0.666/0.003       78,262,436
 4     32.989%      2.618/2.639/2.649/0.007        0.637/0.648/0.654/0.005       76,616,183
 5     32.483%      2.781/2.803/2.815/0.009        0.638/0.644/0.649/0.004       75,441,984
 6     32.318%      3.392/3.414/3.425/0.009        0.632/0.640/0.645/0.004       75,058,801
 7     32.061%      4.654/4.675/4.688/0.010        0.633/0.644/0.650/0.005       74,461,300
 8     31.978%      7.149/7.192/7.212/0.014        0.635/0.643/0.650/0.005       74,269,032
 9     31.961%      9.729/9.761/9.784/0.016        0.638/0.645/0.650/0.003       74,230,342

 avg1  34.188%                        3.949                          0.654
 tot                               1066.332                        176.447      714,603,316
```
### ZLIB-NG (With inflate changes)
```
 Tool: minigzip-ng-infcrc.exe Levels: 1-9
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     44.807%      1.149/1.166/1.177/0.007        0.550/0.558/0.566/0.004      104,063,993
 2     35.391%      1.805/1.828/1.838/0.010        0.553/0.563/0.569/0.004       82,195,416
 3     33.698%      2.438/2.456/2.469/0.009        0.534/0.543/0.550/0.004       78,262,436
 4     32.989%      2.732/2.757/2.771/0.010        0.522/0.527/0.532/0.003       76,616,183
 5     32.484%      2.905/2.926/2.948/0.013        0.513/0.523/0.529/0.004       75,443,521
 6     32.318%      3.515/3.530/3.540/0.007        0.511/0.520/0.525/0.004       75,057,363
 7     32.061%      4.737/4.755/4.768/0.008        0.511/0.523/0.529/0.005       74,461,300
 8     31.978%      7.241/7.264/7.279/0.010        0.513/0.521/0.527/0.004       74,269,032
 9     31.961%      9.807/9.835/9.849/0.011        0.514/0.523/0.531/0.005       74,230,342

 avg1  34.187%                        4.057                          0.534
 tot                               1095.468                        144.062      714,599,586
```
Shows it is possible to get %18 improvement in decompression.

jtkukunas committed:
Reorganize inflate window layout

This commit significantly improves inflate performance by reorganizing the window buffer into a contiguous window and pending output buffer. The goal of this layout is to reduce branching, improve cache locality, and enable for the use of crc folding with gzip input.

The window buffer is allocated as a multiple of the user-selected window size. In this commit, a factor of 2 is utilized.

The layout of the window buffer is divided into two sections. The first section, window offset [0, wsize), is reserved for history that has already been output. The second section, window offset [wsize, 2 * wsize), is reserved for buffering pending output that hasn't been flushed to the user's output buffer yet.

The history section grows downwards, towards the window offset of 0. The pending output section grows upwards, towards the end of the buffer. As a result, all of the possible distance/length data that may need to be copied is contiguous. This removes the need to stitch together output from 2 separate buffers.

In the case of gzip input, crc folding is used to copy the pending output to the user's buffers.

Co-authored-by: Nathan Moinvaziri <nathan@nathanm.com>

nmoinvaz commented:
Ignore the first and second commit please.

mtl1979 commented on inffast.c:
```diff
@@ -127,36 +122,25 @@ void Z_INTERNAL zng_inflate_fast(PREFIX3(stream) *strm, unsigned long start) {
-    beg = out - (start - strm->avail_out);
-    end = out + (strm->avail_out - (INFLATE_FAST_MIN_LEFT - 1));
-    safe = out + strm->avail_out;
+    wsize = state->wsize;
+    out = state->window + wsize + state->wnext;
+    end = state->window + (wsize * 2) - STD_MAX_MATCH;
+    safe = end - state->chunksize;
```
This doesn't look right... `out` and `safe` don't make any sense...

nmoinvaz replied:
`out` makes sense to me, it is writing to the second part of the `state->window` memory allocation. So at `wsize` (window size) + `state->wnext` which is number of bytes written out.

One thing I have thought of was using a `s->pending` similar to what deflate uses, so the calculation `state->window + wsize` doesn't have to keep being done, however it changes more lines of code, so I thought I would leave that change for another PR if we wanted to do something like that.

mtl1979 replied:
Recalculating first part of `out` (`state->window + wsize`) every time is kinda pointless. The original code didn't do that.

I'm not sure what's the point of subtracting `state->chunksize` from `safe` as it is already subtracting `STD_MAX_MATCH` which is a lot more. Those guard bytes should be in addition to the window size, not included in it.

nmoinvaz replied:
Great point. Let me look into it.

nmoinvaz replied:
Ok I have removed the `safe` variable altogether in my most recent commit. Good find.

mtl1979 commented:
When state->wnext is `0`, it will write to the guard bytes instead of skipping them... Basically when we added the vector optimized versions, the end of window had guard bytes, so we could read at least chunk size of bytes without getting random bytes or access violation if the page is unreadable/unwritable. When calculating the total size, either we limit the guard bytes to exactly chunk size, or we distribute the guard bytes 50%/50% between window and output buffer, and end of output buffer.

We could also cache the start of output buffer, like it was done before this pull request to make it clean and easy to understand.

nmoinvaz commented:
I'm not sure we need guard bytes at the end of the window portion of the buffer. I think it is okay to only have guard bytes at the end of the memory allocation. I think we only need the guard bytes for the chunksize operations. But chunksize operations only occur when writing to the output buffer. When writing to the window portion of the buffer, it uses either `memcpy` or `crc32_fold_copy`. Both which *i believe* don't write past the end of where you expect. What do you think?

> We could also cache the start of output buffer, like it was done before this pull request to make it clean and easy to understand.

Can you point me to what you are talking about? I'm unfamilar with it.

nmoinvaz commented on inflate.c:
```diff
@@ -914,69 +859,73 @@ int32_t Z_EXPORT PREFIX(inflate)(PREFIX3(stream) *strm, int32_t flush) {
+                goto inf_leave;
+            if (state->offset > state->wnext + state->whave) {
+                if (state->sane) {
+                    SET_BAD("invalid distance too far back");
```
This line is hit during libpng failed tests.

mtl1979 replied:
Printing out the actual values would give a clue what goes wrong... It could be that either `wnext` or `whave` is not correctly updated. I'm pretty confident that `offset` is always correct, otherwise it would have failed with the original code.

nmoinvaz replied:
* The `wsize` is 256 because `wbits` was 8. 
* `left` variable was set to `strm->avail_out`. That was okay before, when data was copying directly to `strm->next` but now it is copied to internal buffer of size `wsize`. So now we much check to see that `left` is not greater than `strm->wsize`.
* Even with that change though, `wnext = 2, whave = 256, offset = 513, copy = buf_left = 254`. Am I correct in thinking `offset` points to 256 + 256 = 512 which would be past the end of the pending output portion of the memory allocation?
* How can `wsize = 256` when max match length is 258?

nmoinvaz replied:
I think perhaps the errors from libpng are expected. 
```
inflate: allocated
inflate: reset
inflate:   zlib header ok
inflate:     dynamic codes block (last)
inflate:       table sizes ok
inflate:       code lengths ok
inflate:       codes ok
inflate:         literal 'D'
inflate:         literal 'e'
inflate:         literal 'd'
inflate:         literal 'i'
inflate:         literal 'c'
inflate:         literal 'a'
inflate:         literal 't'
inflate:         literal 'e'
inflate:         literal 'd'
inflate:         literal ' '
inflate:         literal 't'
inflate:         literal 'o'
inflate:         literal ' '
inflate:         literal 't'
inflate:         literal 'h'
inflate:         literal 'e'
inflate:         literal ' '
inflate:         literal 'p'
inflate:         literal 'u'
inflate:         literal 'b'
inflate:         literal 'l'
inflate:         literal 'i'
inflate:         literal 'c'
inflate:         literal ' '
inflate:         literal 'd'
inflate:         literal 'o'
inflate:         literal 'm'
inflate:         literal 'a'
inflate:         literal 'i'
inflate:         literal 'n'
inflate:         literal ' '
inflate:         literal 'p'
inflate:         literal 'e'
inflate:         literal 'r'
inflate:         literal ' '
inflate:         literal 'C'
inflate:         literal 'r'
inflate:         literal 'e'
inflate:         literal 'a'
inflate:         literal 't'
inflate:         literal 'i'
inflate:         literal 'v'
inflate:         literal 'e'
inflate:         literal ' '
inflate:         literal 'C'
inflate:         literal 'o'
inflate:         literal 'm'
inflate:         literal 'm'
inflate:         literal 'o'
inflate:         literal 'n'
inflate:         literal 's'
inflate:         literal ' '
inflate:         length 3
inflate:         distance 32
inflate:         literal 'e'
inflate:         literal 'n'
inflate:         literal 's'
inflate:         literal 'e'
inflate:         literal ' '
inflate:         literal '"'
inflate:         literal 'C'
inflate:         literal 'C'
inflate:         literal '0'
inflate:         literal ' '
inflate:         literal '1'
inflate:         literal '.'
inflate:         literal '0'
inflate:         literal '"'
inflate:         literal ';'
inflate:         literal ' '
inflate:         literal 'h'
inflate:         literal 't'
inflate:         literal 't'
inflate:         literal 'p'
inflate:         literal 's'
inflate:         literal ':'
inflate:         literal '/'
inflate:         literal '/'
inflate:         literal 'c'
inflate:         length 7
inflate:         distance 44
inflate:         literal 'c'
inflate:         length 6
inflate:         distance 43
inflate:         literal '.'
inflate:         literal 'o'
inflate:         literal 'r'
inflate:         literal 'g'
inflate:         literal '/'
inflate:         length 6
inflate:         distance 82
inflate:         length 6
inflate:         distance 81
inflate:         literal '/'
inflate:         literal 'z'
inflate:         literal 'e'
inflate:         literal 'r'
inflate:         literal 'o'
inflate:         literal '/'
inflate:         length 3
inflate:         distance 52
inflate:         literal '/'
inflate:         end of block
inflate:   check matches trailer
inflate: reset
inflate:   zlib header ok
inflate:     dynamic codes block (last)
inflate:       table sizes ok
inflate:       code lengths ok
inflate:       codes ok
inflate:         literal 'D'
inflate:         literal 'e'
inflate:         literal 'd'
inflate:         literal 'i'
inflate:         literal 'c'
inflate:         literal 'a'
inflate:         literal 't'
inflate:         literal 'e'
inflate:         literal 'd'
inflate:         literal ' '
inflate:         literal 't'
inflate:         literal 'o'
inflate:         literal ' '
inflate:         literal 't'
inflate:         literal 'h'
inflate:         literal 'e'
inflate:         literal ' '
inflate:         literal 'p'
inflate:         literal 'u'
inflate:         literal 'b'
inflate:         literal 'l'
inflate:         literal 'i'
inflate:         literal 'c'
inflate:         literal ' '
inflate:         literal 'd'
inflate:         literal 'o'
inflate:         literal 'm'
inflate:         literal 'a'
inflate:         literal 'i'
inflate:         literal 'n'
inflate:         literal ' '
inflate:         literal 'p'
inflate:         literal 'e'
inflate:         literal 'r'
inflate:         literal ' '
inflate:         literal 'C'
inflate:         literal 'r'
inflate:         literal 'e'
inflate:         literal 'a'
inflate:         literal 't'
inflate:         literal 'i'
inflate:         literal 'v'
inflate:         literal 'e'
inflate:         literal ' '
inflate:         literal 'C'
inflate:         literal 'o'
inflate:         literal 'm'
inflate:         literal 'm'
inflate:         literal 'o'
inflate:         literal 'n'
inflate:         literal 's'
inflate:         literal ' '
inflate:         length 3
inflate:         distance 32
inflate:         literal 'e'
inflate:         literal 'n'
inflate:         literal 's'
inflate:         literal 'e'
inflate:         literal ' '
inflate:         literal '"'
inflate:         literal 'C'
inflate:         literal 'C'
inflate:         literal '0'
inflate:         literal ' '
inflate:         literal '1'
inflate:         literal '.'
inflate:         literal '0'
inflate:         literal '"'
inflate:         literal ';'
inflate:         literal ' '
inflate:         literal 'h'
inflate:         literal 't'
inflate:         literal 't'
inflate:         literal 'p'
inflate:         literal 's'
inflate:         literal ':'
inflate:         literal '/'
inflate:         literal '/'
inflate:         literal 'c'
inflate:         length 7
inflate:         distance 44
inflate:         literal 'c'
inflate:         length 6
inflate:         distance 43
inflate:         literal '.'
inflate:         literal 'o'
inflate:         literal 'r'
inflate:         literal 'g'
inflate:         literal '/'
inflate:         length 6
inflate:         distance 82
inflate:         length 6
inflate:         distance 81
inflate:         literal '/'
inflate:         literal 'z'
inflate:         literal 'e'
inflate:         literal 'r'
inflate:         literal 'o'
inflate:         literal '/'
inflate:         length 3
inflate:         distance 52
inflate:         literal '/'
inflate:         end of block
inflate:   check matches trailer
inflate: reset
inflate:   zlib header ok
inflate:     dynamic codes block (last)
inflate:       table sizes ok
inflate:       code lengths ok
inflate:       codes ok
inflate:         literal 0x01
inflate:         literal 0x00
inflate:         literal 0x00
inflate:         literal 0x00
inflate:         literal 0x00
inflate:         literal 0x01
inflate:         length 258
inflate:         distance 1
inflate:         length 258
inflate:         distance 1
inflate:         length 258
inflate:         distance 1
inflate:         length 245
inflate:         distance 1
inflate:         literal 0x02
inflate:         literal 0x01
inflate:         literal 0x03
inflate:         literal 0x05
inflate:         literal 0x07
inflate:         length 258
inflate:         distance 4
inflate:         length 258
inflate:         distance 4
inflate:         length 258
inflate:         distance 4
inflate:         length 246
inflate:         distance 4
inflate:         literal 0x02
inflate:         literal 0x00
inflate:         literal 0x02
inflate:         literal 0x04
inflate:         literal 0x06
inflate:         length 258
inflate:         distance 4
inflate:         length 258
inflate:         distance 4
inflate:         length 258
inflate:         distance 4
inflate:         length 246
inflate:         distance 4
inflate:         length 258
inflate:         distance 1025
```

Here distance cannot be 1025 because `window (256) + output buffer (256) = 512`

nmoinvaz replied:
Actually, the `wbits` are different each time the stream is reset. I have made a commit which should fix it.

mtl1979 replied:
So only way to fix the issue is to make sure inflate allocates larger buffer than twice the window size. Basically we would need 2048 bytes or more so it would still be 2^N.

nmoinvaz replied:
The problem is that `wsize` was never set to 0 after window freed (or before window needed to be reallocated). This caused it to have a non-zero value when it goes to `inflate_ensure_window` and thus never gets set to the proper value.

mtl1979 replied:
Doesn't really matter why the window size was wrong, this is still a draft so it is supposed to have bugs... Only way to get the bugs fixed is to understand why something doesn't work correctly.

mtl1979 commented on inflate.c:
```diff
@@ -1248,7 +1213,8 @@ int32_t Z_EXPORT PREFIX(inflateCopy)(PREFIX3(stream) *dest, PREFIX3(stream) *sou
     if (state->window != NULL) {
-        window = (unsigned char *)ZALLOC_WINDOW(source, 1U << state->wbits, sizeof(unsigned char));
+        wsize = 1U << state->wbits;
+        window = (unsigned char *)ZALLOC_WINDOW(state->strm, wsize + state->chunksize, 2);
```
This allocates `(wsize + state->chunksize) * 2)` when we only need `(wsize * 2) + state->chunksize`... I know it's just few bytes, but when `chunksize` gets bigger with AVX512 and later, it will be quite a lot of extra bytes.

nmoinvaz replied:
Fixed. I think that `deflate.c @ line 253 and 1033` also suffers from the same problem.

nmoinvaz replied:
Fixed.

mtl1979 commented on inflate.c:
```diff
@@ -1266,7 +1232,7 @@ int32_t Z_EXPORT PREFIX(inflateCopy)(PREFIX3(stream) *dest, PREFIX3(stream) *sou
     if (window != NULL) {
         wsize = 1U << state->wbits;
-        memcpy(window, state->window, wsize);
+        memcpy(window, state->window, (wsize + state->chunksize) * 2);
```
Like above... Only `wsize` needs to be multiplied by `2`.

nmoinvaz replied:
Fixed.

nmoinvaz commented:
_Ubuntu Reduced Memory_ CI errors with `incorrect data check`.  Tried to reproduce with artifact files from the CI run, but they don't reproduce on Windows.

nmoinvaz commented:
Latest benchmark:
### ZLIB-NG https://github.com/zlib-ng/zlib-ng/commit/e128537d4db2a4bec431d75b92b7b8e01fbf8bd7
```
 Tool: ../develop/Release/minigzip.exe Levels: 0-9       
 Runs: 120        Trim worst: 62        

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.151/0.152/0.153/0.001        0.285/0.287/0.288/0.001      212,804,994
 1     44.256%      0.953/0.958/0.960/0.002        0.604/0.606/0.607/0.001       94,170,816
 2     35.400%      1.586/1.587/1.589/0.001        0.623/0.625/0.627/0.001       75,326,570
 3     33.767%      2.166/2.170/2.171/0.001        0.603/0.607/0.611/0.003       71,853,348
 4     33.061%      2.430/2.433/2.435/0.001        0.594/0.597/0.599/0.001       70,350,209
 5     32.544%      2.541/2.546/2.549/0.002        0.591/0.595/0.600/0.002       69,250,489
 6     32.393%      2.982/2.988/2.991/0.002        0.590/0.593/0.597/0.002       68,928,961
 7     32.137%      4.046/4.051/4.054/0.002        0.591/0.594/0.596/0.001       68,384,900
 8     32.052%      6.348/6.354/6.358/0.003        0.595/0.598/0.599/0.001       68,202,360
 9     31.770%      6.741/6.744/6.748/0.002        0.590/0.591/0.593/0.001       67,603,299

 avg1  40.739%                        2.998                          0.569
 avg2  45.265%                        3.331                          0.633
 tot                               1739.039                        330.180      866,875,946
```
### PR https://github.com/nmoinvaz/zlib-ng/commit/f7e19db6caaa7d9c0513f66ae382fbbac3ccdfe2
```
 Tool: ../push/Release/minigzip.exe Levels: 0-9       
 Runs: 120        Trim worst: 62        

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.151/0.152/0.153/0.001        0.147/0.148/0.150/0.001      212,804,994
 1     44.256%      0.949/0.952/0.955/0.002        0.466/0.467/0.469/0.001       94,170,816
 2     35.400%      1.591/1.593/1.594/0.001        0.484/0.485/0.486/0.001       75,326,570
 3     33.767%      2.164/2.166/2.168/0.001        0.467/0.471/0.476/0.003       71,853,348
 4     33.061%      2.419/2.426/2.429/0.002        0.454/0.458/0.460/0.002       70,350,209
 5     32.544%      2.541/2.545/2.548/0.002        0.451/0.453/0.454/0.001       69,250,489
 6     32.393%      2.984/2.988/2.990/0.001        0.448/0.453/0.456/0.003       68,928,961
 7     32.137%      4.041/4.046/4.049/0.002        0.450/0.454/0.455/0.001       68,384,900
 8     32.052%      6.337/6.346/6.349/0.003        0.455/0.457/0.458/0.001       68,202,360
 9     31.770%      6.763/6.767/6.770/0.002        0.449/0.450/0.452/0.001       67,603,299

 avg1  40.739%                        2.998                          0.430
 avg2  45.265%                        3.331                          0.477
 tot                               1738.867                        249.176      866,875,946
```
In this performance test it appears to be ~24.5% faster inflate. It is different than my initial test done in the description of this issue, because the data sets are different.

nmoinvaz commented:
Unable to reproduce _Reduced Memory CI_ issue with running CTest or reproducing via the artifacts files on Ubuntu 20 with Clang 10.

nmoinvaz commented:
I figured out how to reproduce it. It should be fixed now.

nmoinvaz commented:
Any suggestions on how to make `inflateBack` use `inflate_fast` again?

mtl1979 commented:
@nmoinvaz If inflateBack would just use the "new" inflate_fast directly, it would error out if the match is more than halfway the total size of allocated window as half of the "preallocated window" would essentially be used for double buffering. I was thinking it might be best solution to have "local copy" of the "original" inflate_fast implementation in infback.c. Even if that increases the binary size, it might be the only way to avoid speed penalty in inflateBack.

nmoinvaz commented:
> I was thinking it might be best solution to have "local copy" of the "original" inflate_fast implementation in infback.c. Even if that increases the binary size, it might be the only way to avoid speed penalty in inflateBack.

I was thinking something similar. Maybe use a template (inffast_tpl.c or something). Or we can have one inflate_fast function that has a couple of branches depending on which way it should process the matches and handle the output but it might get messy.

mtl1979 commented:
@nmoinvaz I was also worried that using a template file could make it messy or hard to read... But there might be some code that can still be reused. Using branches and single copy of inflate_fast function might make it a lot slower than using two "customized" copies of the function.

nmoinvaz committed:
Add back original version of inflate_fast for use with inflateBack.

nmoinvaz commented:
This PR is ready for review. Obviously the final PR will have the CI changes removed.

nmoinvaz commented:
Rebased.

Dead2 commented:
Sorry about being slow to review this, but as mentioned, big changes at work, plus I went on a weeks vacation and came back with a bad flu. First day of "recovering" today. But currently none of my usual test-systems are online. So I'll try to set up some machine or another with a reliable testing configuration so that I can do some in-depth benchmarking and testing soon™️.

It looks awesome though, I am seriously impressed. 🏆 

It does need a bit of a rebase now though, so please do that whenever you have the time.

nmoinvaz commented:
@Dead2 thanks I hope it passes the benchmark tests. I am still not doing well. I can rebase at a later time.

nmoinvaz commented:
Rebased. However some unit tests fail due to `test_dict_inflate` in `example`. 

![image](https://user-images.githubusercontent.com/1364432/138539788-c744785a-4f84-43a0-a64e-a8897c403315.png)

`check_dictionary_length = 20` and `sizeof(dictionary) = 6`.

Seems like these checks I added in:
https://github.com/zlib-ng/zlib-ng/pull/1026/files

But perhaps they are not correct in the first place because they had to be disabled for IBM Z. @iii-i

nmoinvaz commented:
Noticed this in IBM Z commit:
> IBM Z: Do not check inflateGetDictionary() with DFLTCC
The zlib manual does not specify a strict contract for
inflateGetDictionary(), it merely says that it "Returns the sliding
dictionary being maintained by inflate", which is an implementation
detail. IBM Z inflate's behavior differs from that of software, and
may change in the future to boot.

nmoinvaz commented:
@iii-i for some reason the CI for S390X fails recently, can you take a look?

iii-i commented:
Currently DFLTCC code relies on software window layout being the same as the hardware one - I'll check what can be done about this.

iii-i committed:
DFLTCC update for window optimization from Jim & Nathan

Stop relying on software and hardware inflate window formats being the
same and act the way we already do for deflate: provide and implement
window-related hooks.

Another possibility would be to use an in-line history buffer (by not
setting HBT_CIRCULAR), but this would require an extra memmove().

Also fix a couple corner cases in the software implementation of
inflateGetDictionary() and inflateSetDictionary().

iii-i commented:
I've pushed https://github.com/iii-i/zlib-ng/commit/5427f308dcb874b95d06be8191f3b4dd97582b18 - please have a look.

I think it might be worth considering restoring the `inflateGetDictionary()` checks - I found a couple logic bugs in that area (my commit fixes them as well).

nmoinvaz commented:
It seems only the `inflateGetDictionary` check for the dictionary length fails. It still returns 20 instead of 6. So somewhere, `state->whave` is getting updated in the second call to `inflate`. The `state->whave` get changed in `window_output_flush` because `state->wnext` is 14. And `state->wnext` gets changed at the end of `zng_inflate_fast`. So if it is using `state->wnext` then it technically isn't the length of the dictionary..? Perhaps it just stores the dictionary in the window and that is overwritten?

The other `inflateGetDictionary` check now succeeds which is excellent.

iii-i commented:
@nmoinvaz Actually 20 vs 6 isn't too bad - this is what I get with DFLTCC as well. This makes perfect sense to me: `Returns the sliding dictionary being maintained by inflate` must refer to the 32K window, so it should contain the initial dictionary + what has been extracted so far (or some subset of it, in case e.g. >32K data has been decompressed).

nmoinvaz commented:
Makes sense to me. I will update the test case to reflect.

nmoinvaz committed:
Fixed inflateGetDictionary length check may include bytes added by last call to inflate.

nmoinvaz commented:
@iii-i Looks like S390X CI jobs still failing due to `example_dict_fuzzer` CTest run.

iii-i commented:
`dfltcc_inflate()` assumed that DFLTCC and software window formats match. Please try the updated fix:

https://github.com/iii-i/zlib-ng/commit/25706250457cdd07af7bd06d17656833e18e4754

Btw, this part in the fuzzer looks scary:

```
    /* deflateBound does not provide enough space for low compression levels. */
    *comprLen = 100 + 2 * PREFIX(deflateBound)(&c_stream, (unsigned long)dataLen);
    *compr = (uint8_t *)calloc(1, *comprLen);
```

Do we know what is the problem there?

nmoinvaz commented:
@iii-i that could be due to Intel quick algorithm. I know we had to make some changes for it to `compressBound`.

nmoinvaz commented:
@iii-i looks like that last commit fixed the issue.

nmoinvaz commented:
Rebased.

Dead2 commented:
aarch64 tests.

### Baseline e12fc9da60ae14969fb6853e51c070739f03f1b5
```
 Tool: minigzip   Levels: 1-9
 Runs: 40         Trim worst: 20

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.166%      2.770/2.802/2.823/0.012        0.967/1.003/1.022/0.016       42,618,858
 2     43.879%      2.779/2.837/2.873/0.026        0.587/0.614/0.627/0.011       20,714,938
 3     42.383%      3.659/3.722/3.742/0.023        0.541/0.575/0.604/0.016       20,008,755
 4     41.660%      2.753/2.777/2.793/0.013        0.359/0.380/0.399/0.013       13,111,391
 5     41.216%      2.980/3.017/3.035/0.017        0.351/0.369/0.383/0.011       12,971,746
 6     41.039%      3.488/3.520/3.540/0.016        0.348/0.370/0.382/0.009       12,916,025
 7     40.778%      2.391/2.411/2.422/0.010        0.161/0.179/0.186/0.007        6,416,924
 8     40.704%      2.928/2.945/2.954/0.007        0.161/0.176/0.183/0.006        6,405,244
 9     40.409%      3.359/3.376/3.387/0.009        0.171/0.189/0.203/0.011        6,358,951

 avg1  42.915%                        3.045                          0.428
 tot                                548.129                         77.113      141,522,832
```

### PR #1033 4988b1f5b274f0d28681e047cb2839b33024237e
```
 Tool: minigzip   Levels: 1-9
 Runs: 40         Trim worst: 20

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.166%      2.765/2.794/2.815/0.016        0.956/0.978/1.000/0.015       42,618,858
 2     43.879%      2.722/2.818/2.856/0.036        0.527/0.590/0.611/0.019       20,714,938
 3     42.383%      3.672/3.725/3.749/0.021        0.551/0.573/0.589/0.010       20,008,755
 4     41.660%      2.752/2.783/2.806/0.017        0.338/0.367/0.382/0.011       13,111,391
 5     41.216%      2.997/3.025/3.039/0.011        0.339/0.358/0.380/0.012       12,971,746
 6     41.039%      3.510/3.536/3.559/0.016        0.353/0.367/0.377/0.007       12,916,025
 7     40.778%      2.386/2.414/2.427/0.011        0.156/0.171/0.182/0.007        6,416,924
 8     40.704%      2.907/2.942/2.952/0.011        0.156/0.170/0.184/0.009        6,405,244
 9     40.409%      3.357/3.380/3.393/0.010        0.156/0.182/0.193/0.011        6,358,951

 avg1  42.915%                        3.046                          0.417
 tot                                548.342                         75.106      141,522,832
```

Decompression is ~2.67% faster.

nmoinvaz commented:
I'm surprised on ARM the gains are not more. Perhaps it is the crc folding that really makes the difference.

mtl1979 commented on arch/s390/dfltcc_detail.h:
```diff
@@ -197,3 +197,50 @@ struct dfltcc_state {
+        param->ho = (param->ho + (n - HB_SIZE)) % HB_SIZE;
+        param->hl = HB_SIZE;
+    }
+}
```
Missing empty line at end of file.

nmoinvaz replied:
Fixed.

mtl1979 commented on arch/s390/dfltcc_inflate.c:
```diff
@@ -133,5 +131,30 @@ int Z_INTERNAL dfltcc_inflate_disable(PREFIX3(streamp) strm) {
     return 0;
 }
+
+int Z_INTERNAL dfltcc_inflate_set_dictionary(PREFIX3(streamp) strm,
```
Too short line...

nmoinvaz replied:
Fixed.

mtl1979 commented on arch/s390/dfltcc_inflate.h:
```diff
@@ -12,6 +12,9 @@ typedef enum {
 dfltcc_inflate_action Z_INTERNAL dfltcc_inflate(PREFIX3(streamp) strm, int flush, int *ret);
 int Z_INTERNAL dfltcc_was_inflate_used(PREFIX3(streamp) strm);
 int Z_INTERNAL dfltcc_inflate_disable(PREFIX3(streamp) strm);
+int Z_INTERNAL dfltcc_inflate_set_dictionary(PREFIX3(streamp) strm,
```
Too short line...

nmoinvaz replied:
Fixed.

mtl1979 commented on test/infcover.c:
```diff
@@ -293,6 +288,11 @@ static void inf(char *hex, char *what, unsigned step, int win, unsigned len, int
+
+    mem_limit(&strm, 1);
+    ret = PREFIX(inflateInit2)(&strm, win);
+                                                assert(ret == Z_MEM_ERROR);
```
Bad indentation.

nmoinvaz replied:
This is how a lot of the tests in `infcover.c` are. I think it better if those changes go in a separate PR.

mtl1979 replied:
Yeah... But only if the first part of line is past the tab position of assert() function... Otherwise it's all on same line.

nmoinvaz replied:
Ok it is fixed.

Dead2 commented:
@iii-i @nmoinvaz The DFLTCC changes are fixes that could be done separately, correct?
If so, perhaps we could do those in a separate PR, and get this one cleaned up a bit.
Many of the 14 commits in this PR seem to be fixes to itself and could be squashed into their parent commits.

nmoinvaz commented on inflate_p.h:
```diff
@@ -99,3 +149,62 @@
+
+static inline void inf_crc_copy(PREFIX3(stream) *strm, unsigned char *const dst,
+        const unsigned char *const src, size_t len)
+{
```
Put on same line.

nmoinvaz replied:
Done.

nmoinvaz commented on inflate_p.h:
```diff
@@ -99,3 +149,62 @@
+}
+
+static inline void window_output_flush(PREFIX3(stream) *strm)
+{
```
Put on same line.

nmoinvaz replied:
Done.

nmoinvaz commented:
@Dead2 from what I understand the DFLTCC changes are specifically for this PR. The CI was failing without the changes.

You are right that a lot of these commits can be squashed. I will try and clean it up a bit. I don't know that there will be an easy way to review this PR.

nmoinvaz commented:
@Dead2 I have squashed some related commits.

iii-i commented:
@Dead2, @nmoinvaz is right - DFLTCC changes are specific to this PR and are not needed without it.
